### ISSUE TITLE
cloud vmware get_instance [size]

### ISSUE THREAD

**Issue by steverweber** [saltstack/salt#38814]
for vmware cloud.get_instance lists good data about a vm.
```salt-call cloud.get_instance jupyter-phys239.uwaterloo.ca```

most of the data is in a nice dict format... however 'size' is a string that looks like a dict.
```
...
    private_ips:
        - 129.97.206.30
        - fe80::250:56ff:febc:72c5
        - 172.17.0.1
    public_ips:
    size:
        cpu: 4
        ram: 8192 MB
...
```
this caused me some pain.
`size = u"cpu: {0}\nram: {1}".format(cpu, ram)`

this is frustrating because I need to split the string up to find the cpu core count and memory count.

https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/vmware.py#L982
```
    if 'size' in selection:
        cpu = vm["config.hardware.numCPU"] if "config.hardware.numCPU" in vm else "N/A"
        ram = "{0} MB".format(vm["config.hardware.memoryMB"]) if "config.hardware.memoryMB" in vm else "N/A"
        vm_select_info['size'] = u"cpu: {0}\nram: {1}".format(cpu, ram)
```

would be improved in my case if it were,
```
    if 'size' in selection:
        cpu = vm["config.hardware.numCPU"] if "config.hardware.numCPU" in vm else "N/A"
        ram = "{0} MB".format(vm["config.hardware.memoryMB"]) if "config.hardware.memoryMB" in vm else "N/A"
        vm_select_info['size'] =  {'cpu': cpu, 'ram': ram}
```

**Comment by steverweber** [comment]
`size` seems like something that is just a simple string in some other places aswell
https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/vmware.py#L1172
https://github.com/saltstack/salt/blob/develop/salt/cloud/clouds/vmware.py#L1576

would much rather a dict.

**Comment by Ch3LL** [comment]
@steverweber I think if we approach this by adding it into develop and documenting it this could be a possible addition. Agreed adding as dict would be a better arpproach..

When raw outputter I get this: `'size': u'cpu: 2\nram: 8192 MB'` with a dict it would def make more sense. Will approve as a feature request. Thanks!

**Comment by nmadhok** [comment]
This will break functionality for other users who have been using size and expect it to return a string. The only reason it's a string and not a dictionary was to match with other cloud providers.

**Comment by cetanu** [comment]
That is all correct. We do want to move away from providers being normalized across all clouds from what I've heard so that would be fine. The only other thing is communicating and/or documenting the change so that people can adjust - if that's how we want to handle it.

**Comment by nmadhok** [comment]
If the idea and decision is to have no standardization between cloud providers, then this change should be fine as long as it follows the deprecation path. The old functionality should still remain atleast for the next 3 major versions. It would be better in that case to just add cpu and memory as additional keys to the top level dict rather than put them under size.

**Comment by cetanu** [comment]
I agree, that's a much more elegant change. I'll add a commit soon

**Comment by cetanu** [comment]
Previous change reverted, cpu and ram added as additional keys

**Comment by nmadhok** [comment]
@techhat @cachedout Did we decide to move away from cloud drivers from being standardized? If so then we also need to change a lot in the VMware cloud driver including the concept of image and size. If that is the case, I'm fine with https://github.com/saltstack/salt/pull/41647 being merged into develop

**Comment by cetanu** [comment]
Some background here: 

https://groups.google.com/forum/#!topic/salt-cloud/8phr9hQIqvo

https://groups.google.com/forum/#!topic/salt-cloud/JVaXz81jyrs

My interpretation might be off, I am human after all. I'm fine with whatever conclusion we get to, as long as it irons out the flaws while keeping all the good.

**Comment by techhat** [comment]
Ping @gtmanfred

**Comment by gtmanfred** [comment]
The goal is not to change the way that the cloud drivers work now.  But to move to modules, the modules do not have to be standardized across clouds.

But the cloud drivers today should remain standardized.

**Comment by nmadhok** [comment]
@gtmanfred That's what I thought. Thanks for clarifying! 

@cetanu @steverweber We could either keep `size` as it is today (string) and simply add `cpu` and `memory` as additional keys:
```
...
    cpu: 4
    memory: 8192 MB
    size:
        cpu: 4
        ram: 8192 MB
...
```

Or we could add `size_dict` as a dictionary and keep `size` as a string until the next 3 releases with a deprecation warning to users:
```
...
    size:
        cpu: 4
        ram: 8192 MB
    size_dict:
        cpu: 4
        memory: 8192 MB
...
```

PS: We should call it `memory` instead of `ram` since that's what is specified in the profile/map (https://docs.saltstack.com/en/develop/topics/cloud/vmware.html#profiles)

**Comment by cetanu** [comment]
I personally like size_dict better since it would be more noticeable alongside the original size key

**Comment by nmadhok** [comment]
I prefer `size_dict` as well. Let's go with that

**Comment by cachedout** [comment]
We have merged #41647 so I believe this is ready to be closed.
